### ISSUE TITLE
Crazy idea: Auto-convert RadixUI from JSX to django-components with LLM?

### ISSUE THREAD

**Issue by EmilStenstrom** [EmilStenstrom/django-components#656]
[RadixUI](https://www.radix-ui.com/themes/docs/overview/getting-started) is a great design system built up with fully designed JSX-components that can be composed into a full interface. It would we very nice to have a set of components that would compose like they do, and have a full system behind it like RadixUI does.

Idea: I wonder if we could use an LLM to convert all of [RadixUI:s components from JSX](https://github.com/radix-ui/themes/tree/main/packages/radix-ui-themes/src/components) to django-components?

**Comment by JuroOravec** [comment]
Unfortunately it won't be that simple :/ This is the same problem I faced with vuetify.

While converting the templating itself could be maybe done with LLM, the biggest challenge is translating the component JS logic. And without it, the components would be much less useful, maybe even stop working.

The RedixUI is built on React. E.g.see [this component ](https://github.com/radix-ui/primitives/blob/main/packages/react/form/src/Form.tsx).

The component above sets up React bindings (`useState`, etc), and renders the template in a single step / single function call. This was the same with vuetify.

For this to be useable with Django components we'd need to do 4 things:

- split the HTML template and component JS logic. So we'd have "headless" components, that don't generate the HTML themselves, but instead are bound to already-existing HTML at runtime, and manage it 

- if we'd want to have the UI library without react, then we'd also need to replace the react API. I did similar for vuetify, where I recreated Vue reactivity API  in AlpineJS, and then in the actual component code I replaced the imports

- another challenge with vuetify was that the whole setup might need to support two kinds of props-passing:
   - passing data down through components in Django/python layer
   - passing data down in JS layer

   The reason for this is that some components may require inputs that are available only in runtime, like a variable holding an HTML element.

   For alpine, I ported Vue's props passing and validation, as well as data emitting behavior, so that the data can still travel up and down the JS components

- lastly, when I used AlpineJS, there was the upside that AlpineJS already offers a way of define components via HTML attributes, using `x-data`. Or some other light HTML manipulations with `x-show`, etc.

   Don't know if there's something similar for React. But if not, then one would need to manage that themselves. In other words, when the page HTML is loaded, some JS script will need to go through the HTML and identify what JS components are declared, and instantiate them. And ideally this would be recursive, so that once a component is initialized, we look into its HTML, and search for nested components, so we can pass the parents data to the child components as props.

And only after all this is done, then the last step would be to generate Django components that render the HTML, whereas the HTML would initialize the runtime components once loaded in browser.

For vuetify it was about 3 weeks of part-time (if you're passionate and know the library) to get to the point where the only thing remaining is the django layer.

The cool thing about this approach tho, is that the same UI component library could then be used in other server-side framework and even other languages. E.g. one could have a Go web server and Go port that would define the same components as the Django one.

I recommend this to be built by someone who would be using this setup right away in their project(s). As there's a lot or things that can go wrong across the whole stack. So the easiest way to debug is if you just NEED the things to work.

**Comment by JuroOravec** [comment]
Ha, nice one!

---

> Thanks for sharing! I'm convinced that a straight conversation isn't a great idea unless you're willing to put in all the necessary work, like you have with vuetify!
> 
> Another approach is to see the list of components as inspiration and try to build a separate set of server-backed versions > of them. That wouldn't lean too heavily on the actual JSX code, but more on the fact that a set of components like that is probably a pretty good view of what components are in most demand.

Yeah, totally can serve as inspo :)
